src/v/cloud_storage/remote_segment.cc

```diff
@@ -50,6 +50,24 @@
 
 #include <exception>
 
+namespace {
+class bounded_stream final : public ss::data_source_impl {
+public:
+    bounded_stream(ss::input_stream<char>& stream, size_t upto)
+      : _stream{stream}
+      , _upto{upto} {}
+
+    ss::future<ss::temporary_buffer<char>> get() override {
+        return _stream.read_up_to(_upto);
+    }
```

Comment:
It actually looks like it can return a shorter buffer, nice catch:

```
    } else if (_buf.size() <= n) {
        // easy case: steal buffer, return to caller
        return make_ready_future<tmp_buf>(std::move(_buf));
    }
```

Fixed

Reply:
I wonder if [read_exactly](https://docs.seastar.io/master/classseastar_1_1input__stream.html#a361edf18947c9a720f2e233043831b79) is a better fit here.